### PR TITLE
Fixed the issue where the function parameter list was not validated when calling an aggregate function with NULL arguments (fix #69195)

### LINKED ISSUE #69195: Unexpected behavior of groupConcat function when using null in expression
https://fiddle.clickhouse.com/6d549ea8-70c0-4d27-a80d-6ad0bee88f9c

```sql
create table t (a Nullable(String)) engine MergeTree order by ();

insert into t values ('123'), ('456'), ('789');

select groupConcat(',', 1, 1, 1, 1)(a, null, a, a, 1, 1) from t;
```

Output:
```
\N
```

Why I do not get an exception executing this invalid query? 

I expect to see either

```
DB::Exception: Incorrect number of parameters for aggregate function groupConcat, should be 0, 1 or 2, got: 5. (TOO_MANY_ARGUMENTS_FOR_FUNCTION)
```
or
```
DB::Exception: Aggregate function groupConcat requires single argument. (NUMBER_OF_ARGUMENTS_DOESNT_MATCH)
```

### PR DESCRIPTION
**Pull request by ligen131**
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fixed the issue where the function parameter list was not validated when calling an aggregate function with NULL arguments (fix #69195)

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by jsc0218** [comment]
@ligen131 thanks for contribution. For the bugfix category, it is necessary to add a test, so CI can verify according to the test.

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 312704a9474c6b1fa10444bd813ff3bee9cf5400 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70674/312704a9474c6b1fa10444bd813ff3bee9cf5400/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/ba0b7f568907bac8c4472bcab5222a49557d1fd4/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by ligen131** [comment]
This bug seems to be much more complex than I initially thought...

In my fix under this PR, I moved the creation of each aggregate function ahead of the NULL arguments check. This is because during the creation of the aggregate functions, the parameter list is checked to ensure it meets expectations, which helps avoid returning NULL results due to the unintended extra inclusion of NULL arguments.

However, after reviewing the failed test cases from the CI, I found that nearly all the aggregate functions have this bug (strictly speaking, all aggregate functions created with `AggregateFunctionProperties properties = { .returns_default_when_only_null = false };`). For example, `avg(1, NULL)` should throw an error `NUMBER_OF_ARGUMENTS_DOESNT_MATCH` instead of returning `\N` or throwing an error `Illegal type Nullable(Nothing) of argument for aggregate function avg. (ILLEGAL_TYPE_OF_ARGUMENT)`, according to issue #69195. There is a potential solution, which involves adding `isNullable()` during the creation and parameter validation process (`allowType`) of all aggregate functions. For instance, for `AggregateFunctionAvg`:

```cpp
bool allowType(const DataTypePtr& type) noexcept
{
    const WhichDataType t(type);
    return t.isInt() || t.isUInt() || t.isFloat() || t.isDecimal() || t.isNullable();
}
```

However, this modification involves significant effort, and I'm not entirely sure if it genuinely aligns with the expected behavior of all aggregate functions. If changes are necessary, I think we need to discuss this further first.

Additionally, I encountered some peculiar phenomena in the test cases. Some functions intentionally included extra NULLs in the parameter list to return NULL instead of throwing an error, such as in `00808_array_enumerate_segfault`:

```sql
SELECT arrayEnumerateDense([STDDEV_SAMP(NULL, 910947.571364)], [NULL]);
```

Theoretically I think, both `STDDEV_SAMP` and `arrayEnumerateDense` should throw a `NUMBER_OF_ARGUMENTS_DOESNT_MATCH` error, but the expected outcome for this test case is a normal return. This makes me uncertain whether #69195 is indeed not the intended behavior.

In summary, I think we need to discuss this issue further.

> For the bugfix category, it is necessary to add a test, so CI can verify according to the test.

BTW, if there is a good solution to this issue, I will add tests for my modifications.

**Comment by jsc0218** [comment]
agree, looks like some tradeoff here, so close this pr.
